In [2]:
# Importamos las librerías necesarias
import pandas as pd
import numpy as np
pd.set_option("display.max_columns", None)
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
# Para imputación de nulos usando métodos estadísticos avanzados
from sklearn.impute import SimpleImputer 
from sklearn.experimental import enable_iterative_imputer # Necesario para IterativeImputer
from sklearn.impute import IterativeImputer 
from sklearn.impute import KNNImputer 

In [4]:
df_hr = pd.read_csv("hr.csv")
df_hr.head(5)

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41.0,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,2,Female,94,3,2,sALES eXECUTIVE,4.0,Single,5993.0,19479,8,Y,Yes,11,3,1,80.0,0,8,0.0,1,6,4,0,5.0
1,49.0,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,3,Male,61,2,2,rESEARCH sCIENTIST,2.0,Married,5130.0,24907,1,Y,No,23,4,4,NaN,1,10,3.0,3,10,7,1,7.0
2,37.0,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,4,Male,92,2,1,lABORATORY tECHNICIAN,3.0,Single,2090.0,2396,6,Y,Yes,15,3,2,NaN,0,7,3.0,3,0,0,0,0.0
3,33.0,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,4,Female,56,3,1,rESEARCH sCIENTIST,3.0,Married,2909.0,23159,1,Y,Yes,11,3,3,80.0,0,8,3.0,3,8,7,3,0.0
4,27.0,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,1,Male,40,3,1,lABORATORY tECHNICIAN,2.0,Married,3468.0,16632,9,Y,No,12,3,4,80.0,1,6,3.0,3,2,2,2,2.0


In [6]:
def exploracion_basica(df):
    print(f"Las dos primeras filas son:")
    display(df.head(2))
    print("=" * 50)
    print(f"Las dos últimas filas son:")
    display(df.tail(2))
    print("=" * 50)
    print(f"Las dos filas aleatorias son:")
    display(df.sample(2)) 
    print("=" * 50)
    print(f"Información sobre DataFrame:")
    print(df.info())
    print("=" * 50)
    print(f"El dataframe tiene {df.shape[0]} filas y {df.shape[1]} columnas.")
    return df

In [7]:
df_hr = exploracion_basica(df_hr)

Las dos primeras filas son:


,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41.0,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,2,Female,94,3,2,sALES eXECUTIVE,4.0,Single,5993.0,19479,8,Y,Yes,11,3,1,80.0,0,8,0.0,1,6,4,0,5.0
1,49.0,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,3,Male,61,2,2,rESEARCH sCIENTIST,2.0,Married,5130.0,24907,1,Y,No,23,4,4,NaN,1,10,3.0,3,10,7,1,7.0


Las dos últimas filas son:


,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
1472,24.0,Yes,Travel_Rarely,240,Human Resources,22,1,Human Resources,1,1714,4,Male,58,1,1,hUMAN rESOURCES,3.0,Married,1555.0,11585,1,Y,No,11,3,3,80.0,1,1,2.0,3,1,0,0,0.0
1473,45.0,No,Travel_Rarely,1339,Research & Development,7,3,Life Sciences,1,86,2,Male,59,3,3,rESEARCH sCIENTIST,1.0,Divorced,9724.0,18787,2,Y,No,17,3,3,NaN,1,25,2.0,3,1,0,0,0.0


Las dos filas aleatorias son:


,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
271,47.0,Yes,Non-Travel,666,Research & Development,29,4,Life Sciences,1,376,1,Male,88,3,3,mANAGER,2.0,Married,11849.0,10268,1,Y,Yes,12,3,4,80.0,1,10,2.0,2,10,7,9,9.0
668,28.0,No,Travel_Rarely,995,Research & Development,9,3,Medical,1,930,3,Female,77,3,1,rESEARCH sCIENTIST,3.0,Divorced,2377.0,9834,5,Y,No,18,3,2,80.0,1,6,2.0,3,2,2,2,2.0


Información sobre DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1474 entries, 0 to 1473
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       1401 non-null   float64
 1   Attrition                 1474 non-null   object 
 2   BusinessTravel            1357 non-null   object 
 3   DailyRate                 1474 non-null   int64  
 4   Department                1445 non-null   object 
 5   DistanceFromHome          1474 non-null   int64  
 6   Education                 1474 non-null   int64  
 7   EducationField            1416 non-null   object 
 8   EmployeeCount             1474 non-null   int64  
 9   EmployeeNumber            1474 non-null   int64  
 10  EnvironmentSatisfaction   1474 non-null   int64  
 11  Gender                    1474 non-null   object 
 12  HourlyRate                1474 non-null   int64  
 13  JobInvolvement            1474 non

In [8]:
df_hr.sample(5)

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
944,28.0,No,Non-Travel,1476,Research & Development,1,3,Life Sciences,1,1315,3,Female,55,1,2,lABORATORY tECHNICIAN,4.0,Married,6674.0,16392,0,Y,No,11,3,1,80.0,3,10,6.0,3,9,8,7,5.0
420,29.0,No,Travel_Rarely,986,Research & Development,3,4,Medical,1,564,2,Male,93,2,3,rESEARCH dIRECTOR,3.0,Married,11935.0,21526,1,Y,No,18,3,3,80.0,0,10,2.0,3,10,2,0,7.0
54,26.0,No,Travel_Rarely,1443,Sales,23,3,Marketing,1,72,3,Female,47,2,2,sALES eXECUTIVE,4.0,Married,4157.0,21436,7,Y,Yes,19,3,3,80.0,1,5,2.0,2,2,2,0,0.0
70,59.0,No,Travel_Frequently,1225,Sales,1,1,Life Sciences,1,91,1,Female,57,2,2,sALES eXECUTIVE,3.0,Single,5473.0,24668,7,Y,No,11,3,4,80.0,0,20,2.0,2,4,3,1,NaN
44,30.0,No,Travel_Frequently,721,Research & Development,1,2,Medical,1,57,3,Female,58,3,2,lABORATORY tECHNICIAN,4.0,Single,4011.0,10781,1,Y,No,23,4,4,80.0,0,12,2.0,3,12,8,3,7.0


**revision para limpieza Maria

- nombre de columnas en minuscula y separado con _
- attrition es object, no boleano. nes que sigue trabajando en la empresa y si que ya no trabaja allí, no?
- education: 1 a 4 siendo 1 nivel de estudios mas bajos y 4 mas alto?
- edad cambiar a int
- emproyee count no se que es
- hourly rate es tarifa por hora.
- job role: cambiar a todo en minusculas
- bussines travel lo dejaria solo con rarely o con frequantly. como deberiamos de ponerlo todo en minusculas no cuesta nada modificarlo apra que quede mas limpio sin que ponga travel
- no entiendo bien monthly income y monthly rate
- departament: poner en minusculas
- hay varias columnas más que no entiendo bien, mejor lo vemos jutnas, porque hay muchas columnas.

* nulos tenemos en 
YearsWithCurrManager      1326 non-null   float64
StandardHours             1310 non-null   float64
OverTime                  1430 non-null   object 
MonthlyIncome             1460 non-null   float6
JobSatisfaction           1445 non-null   float64
MaritalStatus             1342 non-null   object 
EducationField            1416 non-null   object 
Department                1445 non-null   object
BusinessTravel            1357 non-null   object 
Age                       1401 non-null   float64


para ver por que la gente se va se me ocurre que tendriamos que ver 
- cuanta gente se ha ido de la empresa
- si tiene que ver con la antiguedad TotalWorkingYears / YearsAtCompany
- si tiene que ver con el sueldo DailyRate. ver maximo y minimo y donde se situan, o en las otras columnas de sueldo.
- si eran hombres o mujeres
- la edad que tienen
- si viajaban o no
- la distancia al trabajo desde casa
- de que departamento son por si coincide
- si llevaban mucho tiempo con el mismo responsable o si es el mismo responsable el de todos los que se van
- si han tenido aumentos de sueldo
- YearsSinceLastPromotion	
- su grado de sarisfaccion
- su job involvment
- si hacen horas extra
- WorkLifeBalance
- TrainingTimesLastYea
